### ISSUE TITLE
Use other HTTP library in NodeJS API Client 

### ISSUE THREAD

**Issue by thc202** [zaproxy/zaproxy#5969]
The [`request` library is deprecated](https://github.com/request/request#deprecated).

For possible alternatives: request/request#3143

NodeJS API Client repo: https://github.com/zaproxy/zap-api-nodejs

**Comment by lordnodd** [comment]
I went through the possible alternatives, I've used axios and node-fetch in previous projects and since we are using a config for proxy, axios would be a better choice over node-fetch as it supports proxy directly but [this issue](https://github.com/axios/axios/issues/2072#issuecomment-567473812) shows the existing problems and the solution to use http-proxy-agent. 
I'd like to give it a try with axios and http-proxy-agent if package size is not an issue.

**Comment by lordnodd** [comment]
Should I give it a shot till no one shows up @thc202 ?

**Comment by kingthorin** [comment]
I have no objection, however I've never really worked on the Node API Client at all 🤷

**Comment by lordnodd** [comment]
Cool I am on it, by looking at some logs, it looks like only the src/index.js needs to be updated with axios. Rest all the module uses its prototype through ClientApi. I'll try to modify to observe the changes. 

I have a query though, these files are scripts right, they just execute without resulting anything atleast not when running as it is. So how should I test them, to check whether they are working or not.

**Comment by kingthorin** [comment]
@binarymist could you chime in here? Are you good with switching libraries? Do you have testing/usage advice?

**Comment by binarymist** [comment]
Hi there...

Sorry, I didn't realise this issue existed.

This is an [issue](https://github.com/purpleteam-labs/purpleteam/issues/14) in our [backlog](https://github.com/purpleteam-labs/purpleteam/projects/2), which also has a link to a recent upgrade (commit) from the request (and friends) lib(s) to got.

I was planning on using got on zap-api-nodejs. I've used axios on plenty of occasions and in [one place using http-proxy-agent](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/d04c3ebdeb976ea6909392f53df6663b642629ae/src/api/app/models/app.parallel.js#L267-L270). It's currently doing what it should do, but if/when we change the HTTP lib again in the app-scanner, I'd go with got again, mostly due to how one configures it, it's very DRY. Asides: It's also smaller and newer.

Also not sure if you are aware @lordnodd , but you'll need to set-up the Zap Java development environment as well if you haven't already as the zap-api-nodejs lib is generated in Java.

**Comment by lordnodd** [comment]
Oh, Sorry for not replying fast. Would that be much trouble, if I don't know JAVA?

**Comment by kingthorin** [comment]
Shouldn't be a big deal, there's info here to get you started: https://www.zaproxy.org/docs/developer/quick-start-build/

**Comment by binarymist** [comment]
Hi @lordnodd & @kingthorin It's quite a process.

I have a document (that's somewhat specific to an environment I created for this) with all/most of the steps and learnings I went through last time doing a regen. Part time it took several weeks and it was non-trivial. I remember the pain (not trying to put you  off, but rather prepair you if you decide to go for it) My doc has sections on:

* Set-up
* Modifying the Source 
* Generat the additional extensions
* Clean-up
* Build Gotchas (there was a few of these)

FYI: There was Java changes to be done, from memory not that many though. There are quite a few things you need to know with the Dev environment that I guess the Zap Core Team "just knows" that I had to learn (there was some pain here).

There are also other tidy-ups/issues that need to be done when and if you have the environment all set-up such as:

* https://github.com/zaproxy/zaproxy/issues/5126 which links to two other pull requests which there is still work to be done by the look of it
* [Initial NodeJS API Upgrade](https://github.com/zaproxy/zaproxy/issues/5044) on zaproxy. Still todo's in here
  * PR [Node api generation](https://github.com/zaproxy/zaproxy/pull/5088) on zaproxy
  * PR [New api](https://github.com/zaproxy/zap-api-nodejs/pull/1) on zap-api-nodejs
  * PR [Removed superfluous old node api files](https://github.com/zaproxy/zaproxy/pull/5096) on zaproxy
  * PR [Updated node output dir](https://github.com/zaproxy/zap-extensions/pull/1869) on zap-extensions
  * PR [Updated node output dir](https://github.com/zaproxy/zap-extensions/pull/1870) on zap-extensions
  * PR [Updated node output dir](https://github.com/zaproxy/zap-extensions/pull/1871) on zap-extensions
* [Changelog](https://github.com/binarymist/zap-api-nodejs/blob/master/CHANGELOG.md) to be done manually, if it was up to me, I'd be using [np](https://github.com/sindresorhus/np) which publishes automatically and starts the changelog process for you. This would make the release process much easier. I use np on most of my NodeJS projects
* npm audit and what ever else necessary to make sure all dependencies are up to date

There is no point in going through the set-up unless all the tidy-ups get done. Hopefully I haven't put you off yet @lordnodd ?

**Comment by thc202** [comment]
Most of those steps are no longer needed.

**Comment by lordnodd** [comment]
though I feel Java to be overwhelming, thaks for the curated list @binarymist you  and referring to [javadoc](https://www.javadoc.io/doc/org.zaproxy/zap/2.6.0/org/zaproxy/zap/extension/api/CoreAPI.html) made it a little easier, I agree @thc202 I don't think I need to set-up all that(I can be wrong). I have an this much understanding that NODEJSAPIGENERATOR.java is creating those files i see in zap-nodejs-api. File and function generation is automated that way for zaproxy n-features. But how is the library getting require(like `const request = require("request");`)
Is it through the package.json , the way deprecated library like loadash gets updated
OR
Is it from the CORE.java handleApiAction(), i think this is more likely but then how and where does the deprecated request package fit in if its made from java files?

**Comment by njmulsqb** [comment]
Hi,

I would like to take this, seems like changes needs to be done primarily in `NodeJSAPIGenerator.java` and then some in `index.js` to replace request usage with some other lib. 
@binarymist I like your call on using got. I have found this request-to-got migration guide https://github.com/sindresorhus/got/blob/main/documentation/migration-guides/request.md that might come handy.
@lordnodd What sort of challenges you faced when you tried to work on this issue?

**Comment by thc202** [comment]
@trevor-vaughan ping/fyi, since you are using this API client.

**Comment by thc202** [comment]
@njmulsqb check also #7608.

**Comment by trevor-vaughan** [comment]
Considering that the migration is probably a breaking change, are we up for a breaking change fest (as in just migrating to object passing to resolve #7608 )?

Overall, the code is straightforward enough, it'll just take some time to hammer out.

**Comment by thc202** [comment]
That sounds good to me, whatever makes it easier to use (I'm not a user of this API client though).

**Comment by njmulsqb** [comment]
Assign this to me! :)

**Comment by psiinon** [comment]
Assigned 😁

**Comment by njmulsqb** [comment]
I am facing some technical difficulties to debug the JS code, I have set breakpoints in the file, gone to Debug --> Run & Debug, and the debugger opens and closes itself without hitting the breakpoints, I have tried launch configurations as well. 
To see if I am doing it the right way, I created a random JS file and followed the procedure on that, the debugger ran perfectly in VS Code. Can anyone help me with getting the debugger run on nodejs api client, it will come handy to properly migrate the library. Thanks
